File: backend/checker.py
"""
checker.py — search listing text for the word "dishwasher".

Returns a structured result dict that the API layer passes straight back to
the frontend.
"""

from __future__ import annotations

import re

_PATTERN = re.compile(r"dishwasher", re.IGNORECASE)

# How many characters of context to show around the matched word
_CONTEXT_CHARS = 80


def check_text(text: str) -> dict:
    """
    Search *text* for the word "dishwasher" (case-insensitive).

    Returns
    -------
    dict with keys:
        has_dishwasher : bool
        method         : "text"  (always, for this stage)
        evidence       : str | None  — surrounding snippet, or None
    """
    match = _PATTERN.search(text)

    if match:
        start = max(0, match.start() - _CONTEXT_CHARS)
        end = min(len(text), match.end() + _CONTEXT_CHARS)
        snippet = text[start:end].strip()
        # Normalise whitespace so the snippet reads cleanly in the UI
        snippet = re.sub(r"\s+", " ", snippet)
        return {
            "has_dishwasher": True,
            "method": "text",
            "evidence": f"…{snippet}…",
        }

    return {
        "has_dishwasher": False,
        "method": "text",
        "evidence": None,
    }

File: src/lib/checker.js
/**
 * Text-based dishwasher detection.
 * Searches the listing text for the word "dishwasher" (case-insensitive)
 * and returns the surrounding context as evidence.
 *
 * @param {string} text
 * @returns {{ has_dishwasher: boolean, method: 'text', evidence: string|null }}
 */
export function checkText(text) {
  const match = text.match(/dishwasher/i)
  if (!match) {
    return { has_dishwasher: false, method: 'text', evidence: null }
  }

  const idx = match.index
  const start = Math.max(0, idx - 80)
  const end = Math.min(text.length, idx + match[0].length + 80)
  const snippet = text.slice(start, end).replace(/\s+/g, ' ').trim()
  const evidence =
    (start > 0 ? '\u2026' : '') + snippet + (end < text.length ? '\u2026' : '')

  return { has_dishwasher: true, method: 'text', evidence }
}
